src/small.rs

```diff
@@ -484,18 +525,132 @@ impl MPCSmall {
+              vk.alpha_g1
+              vk.beta_g1
+              vk.beta_g2
+              vk.gamma_g2
+              vk.delta_g1
+              vk.delta_g2
+              vk.ic length (4 bytes)
+              vk.ic (G1)
+              h length (4 bytes)
+              h (G1)
+              l length (4 bytes)
+              l (G1)
+              a length (4 bytes)
+              a (G1)
+              b_g1 length (4 bytes)
+              b_g1 (G1)
+              b_g2 length (4 bytes)
+              b_g2 (G2)
+              cs_hash (64 bytes)
+              contributions length (4 bytes)
+              contributions (544 bytes per PublicKey)
+        */
+
+        // Non-raw sizes.
+        let g1_size = G1Affine::uncompressed_size() as u64; // 96 bytes
+        let g2_size = G2Affine::uncompressed_size() as u64; // 192 bytes
+
+        let mut file = File::open(large_path)?;
+
+        // Read delta_g1, delta_g2, and ic's length.
+        let delta_g1_offset = g1_size + g1_size + g2_size + g2_size; // vk.alpha_g1 + vk.beta_g1 + vk.beta_g2 + vk.gamma_g2
+        file.seek(SeekFrom::Start(delta_g1_offset)).unwrap();
+        let delta_g1 = read_g1(&mut file, check_subgroup)?;
+        let delta_g2 = read_g2(&mut file, check_subgroup)?;
+        let ic_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read h's length.
+        let h_len_offset = delta_g1_offset + g1_size + g2_size + 4 + ic_len * g1_size; // + vk.delta_g1 + vk.delta_g2 + ic length + ic
+        file.seek(SeekFrom::Start(h_len_offset)).unwrap();
+        let h_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read l's length.
+        let l_len_offset = h_len_offset + 4 + h_len * g1_size; // + h length + h
+        file.seek(SeekFrom::Start(l_len_offset)).unwrap();
+        let l_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read a's length.
+        let a_len_offset = l_len_offset + 4 + l_len * g1_size; // + l length + l
+        file.seek(SeekFrom::Start(a_len_offset)).unwrap();
+        let a_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read b_g1's length.
+        let b_g1_len_offset = a_len_offset + 4 + a_len * g1_size; // + a length + a
+        file.seek(SeekFrom::Start(b_g1_len_offset)).unwrap();
+        let b_g1_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read b_g2's length.
+        let b_g2_len_offset = b_g1_len_offset + 4 + b_g1_len * g1_size; // + b_g1 length + b_g1
+        file.seek(SeekFrom::Start(b_g2_len_offset)).unwrap();
+        let b_g2_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read cs_hash.
+        let cs_hash_offset = b_g2_len_offset + 4 + b_g2_len * g2_size; // + b_g2 length + b_g2
+        file.seek(SeekFrom::Start(cs_hash_offset)).unwrap();
+        let mut cs_hash = [0u8; 64];
+        file.read_exact(&mut cs_hash)?;
+
+        // Read contribution's length.
+        let contributions_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Reset seek position.
+        drop(file);
```

Comment:
why not use `Seek` to go back to the beginning?

Reply:
A few reasons, but this definitively can be changed
1. this code was written in like a day at the start of the first trusted-setup
2. I don't think `Seek::rewind()` was in std-lib when this code was written (but I could have just done `file.seek(SeekFrom::Start(0))`
3. the set of reads before `drop(file)` use unbuffered-io whereas the reads after `drop(file)` use buffered-io and I just wanted to be sure that I wasn't going to get a bug when creating the buffered reader from a file that was already read from

Reply:
sgtm